### PR TITLE
Adds OOC notes

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request
image

see I'd port skyrat's but I don't like the toggles, they're too blunt for my taste. 

## Why It's Good For The Game

Separate from flavor text, always shows, does not save with DNA.
Don't be that guy who metagames and ruins it for everyone.

## Changelog
:cl:
add: OOC notes has been added for ERP-related preferences. Please do not metagame or abuse this, and it is also not for character descriptions.
/:cl:

### CODE REVIEW

**Comment by silicons** [comment]
oh yeah bug report if you use manage flavor texts in game as opposed to from prefs the window doesn't word-wrap properly but that's not caused by me :^) @Ghommie

**Comment by lolman360** [comment]
kev i thought you didn't use ooc notes

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -38,7 +38,8 @@
 	. = ..()
 	if(!CONFIG_GET(flag/disable_human_mood))
 		AddComponent(/datum/component/mood)
-	AddElement(/datum/element/flavor_text/carbon)
+	AddElement(/datum/element/flavor_text/carbon, _name = "Flavor Text", _save_key = "flavor_text")
+	AddElement(/datum/element/flavor_text/carbon, _name = "OOC Notes", _addendum = "Put information on ERP/vore/lewd-related preferences here. THIS SHOULD NOT CONTAIN REGULAR FLAVORTEXT!!", _always_show = TRUE, _save_key = "ooc_notes", _dna_carryover = FALSE)
```
Comment (Ghommie) [inline code/modules/mob/living/carbon/human/human.dm]: Sorry for the late reply, but just make another `/datum/element/flavor_text` subtype for other prefs stuff or something, `/datum/element/flavor_text/carbon` is for dna stuff.

File: code/datums/elements/flavor_text.dm [hunk 2]
```diff
@@ -5,12 +5,14 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
 	id_arg_index = 3
 	var/flavor_name = "Flavor Text"
 	var/list/texts_by_atom = list()
-	var/addendum = "This can also be used for OOC notes and preferences!"
+	var/addendum = ""
 	var/always_show = FALSE
 	var/max_len = MAX_FLAVOR_LEN
 	var/can_edit = TRUE
+	/// For preference/DNA saving/loading. Null to prevent.
+	var/save_key
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm:13]: Also add a new variable used to override the standard examine list input on lines on lines 64-68.
We need a simple `[OOC]` text href button, without the first 37 characters of possible kinks.

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by silicons** [comment]
moved pref load functionality up, it is only triggered if the prefs have the save key in it.
ooc notes is now base flavor text as it doesn't require any proc overrides

File: code/datums/elements/flavor_text.dm [hunk 3]
```diff
@@ -33,9 +39,12 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
 		LAZYOR(GLOB.mobs_with_editable_flavor_text[M], src)
 		M.verbs |= /mob/proc/manage_flavor_tests
 
+	if(ishuman(target))
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
	if(save_key && ishuman(target))
```

File: code/datums/elements/flavor_text.dm [hunk 4]
```diff
@@ -112,22 +124,28 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
+	if(P.features.Find(save_key))
+		texts_by_atom[H] = P.features[save_key]
+
 //subtypes with additional hooks for DNA and preferences.
 /datum/element/flavor_text/carbon
 	//list of antagonists etcetera that should have nothing to do with people's snowflakes.
 	var/static/list/i_dont_even_know_who_you_are = typecacheof(list(/datum/antagonist/abductor, /datum/antagonist/ert,
 													/datum/antagonist/nukeop, /datum/antagonist/wizard))
+	/// Do we transfer by DNA?
+	var/dna_carryover = TRUE
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
```

File: code/datums/elements/flavor_text.dm [hunk 5]
```diff
@@ -112,22 +124,28 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
 
-/datum/element/flavor_text/carbon/Attach(datum/target, text = "", _name = "Flavor Text", _addendum, _max_len = MAX_FLAVOR_LEN, _always_show = FALSE, _edit = TRUE)
+/datum/element/flavor_text/carbon/Attach(datum/target, text = "", _name = "Flavor Text", _addendum, _max_len = MAX_FLAVOR_LEN, _always_show = FALSE, _edit = TRUE, _save_key = "flavor_text", _examine_no_preview = FALSE, _dna_carryover = TRUE)
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
/datum/element/flavor_text/carbon/Attach(datum/target, text = "", _name = "Flavor Text", _addendum, _max_len = MAX_FLAVOR_LEN, _always_show = FALSE, _edit = TRUE, _save_key = "flavor_text", _examine_no_preview = FALSE)
```

File: code/datums/elements/flavor_text.dm [hunk 6]
```diff
@@ -112,22 +124,28 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
 	if(!iscarbon(target))
 		return ELEMENT_INCOMPATIBLE
 	. = ..()
 	if(. == ELEMENT_INCOMPATIBLE)
 		return
+	src.dna_carryover = _dna_carryover
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
```

File: code/datums/elements/flavor_text.dm [hunk 7]
```diff
@@ -136,15 +154,15 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
 
 /datum/element/flavor_text/carbon/proc/update_dna_flavor_text(mob/living/carbon/C)
-	texts_by_atom[C] = C.dna.features["flavor_text"]
-
-/datum/element/flavor_text/carbon/proc/update_prefs_flavor_text(mob/living/carbon/human/H, datum/preferences/P, icon_updates = TRUE, roundstart_checks = TRUE)
-	texts_by_atom[H] = P.features["flavor_text"]
+	if(!save_key)
+		return
+	// if we aren't carrying over from dna load from prefs instead
+	texts_by_atom[C] = dna_carryover? C.dna.features[save_key] : C.client?.prefs?.features[save_key]
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
	texts_by_atom[C] = C.dna.features[save_key]
```

File: code/datums/elements/flavor_text.dm [hunk 8]
```diff
@@ -136,15 +154,15 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
 
 /datum/element/flavor_text/carbon/set_flavor(mob/living/carbon/user)
 	. = ..()
-	if(. && user.dna)
-		user.dna.features["flavor_text"] = texts_by_atom[user]
+	if(. && user.dna && dna_carryover)		// only save to dna if carrying over.
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
	if(. && user.dna)
```

File: code/datums/elements/flavor_text.dm [hunk 9]
```diff
@@ -136,15 +154,15 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
...
+		user.dna.features["[save_key]"] = texts_by_atom[user]
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
		user.dna.features[save_key] = texts_by_atom[user]
```

File: code/datums/elements/flavor_text.dm [hunk 10]
```diff
@@ -153,4 +171,4 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
 	if(is_type_in_typecache(antag, i_dont_even_know_who_you_are))
 		texts_by_atom[user] = ""
 		if(user.dna)
-			user.dna.features["flavor_text"] = ""
\ No newline at end of file
+			user.dna.features["[save_key]"] = ""
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
			user.dna.features[save_key] = ""
```

File: code/modules/client/preferences.dm [hunk 11]
```diff
@@ -150,6 +150,7 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 		"ipc_screen" = "Sunburst",
 		"ipc_antenna" = "None",
 		"flavor_text" = "",
+		"ooc_notes" = "",
```
Comment (Ghommie) [inline code/modules/client/preferences.dm:153]: Though this is not exactly a dna feature, I'll let it slide.

File: code/modules/client/preferences.dm [hunk 12]
```diff
@@ -345,6 +346,15 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 					dat += "[features["flavor_text"]]"
 			else
 				dat += "[TextPreview(features["flavor_text"])]...<BR>"
+			dat += "<h2>OOC notes</h2>"
+			dat += "<a href='?_src_=prefs;preference=ooc_notes;task=input'><b>Set OOC notes</b></a><br>"
+			if(length(features["ooc_notes"]) <= 40)
+				if(!length(features["ooc_notes"]))
```
Comment (Ghommie) [inline code/modules/client/preferences.dm]:
```suggestion
			var/ooc_notes_len = length(features["ooc_notes"])
			if(ooc_notes_len <= 40)
				if(!ooc_notes_len)
```

**Review by Ghommie — APPROVED** [review]

File: code/datums/elements/flavor_text.dm [hunk 13]
```diff
@@ -136,15 +154,15 @@ GLOBAL_LIST_EMPTY(mobs_with_editable_flavor_text) //et tu, hacky code
 	UnregisterSignal(C, list(COMSIG_CARBON_IDENTITY_TRANSFERRED_TO, COMSIG_MOB_ANTAG_ON_GAIN, COMSIG_HUMAN_PREFS_COPIED_TO, COMSIG_HUMAN_HARDSET_DNA, COMSIG_HUMAN_ON_RANDOMIZE))
 
 /datum/element/flavor_text/carbon/proc/update_dna_flavor_text(mob/living/carbon/C)
-	texts_by_atom[C] = C.dna.features["flavor_text"]
-
-/datum/element/flavor_text/carbon/proc/update_prefs_flavor_text(mob/living/carbon/human/H, datum/preferences/P, icon_updates = TRUE, roundstart_checks = TRUE)
-	texts_by_atom[H] = P.features["flavor_text"]
+	if(!save_key)
+		return
+	// if we aren't carrying over from dna load from prefs instead
```
Comment (Ghommie) [inline code/datums/elements/flavor_text.dm]:
```suggestion
```

**Pull request merged**